### PR TITLE
Patching Factoring - New Abstractions Project

### PR DESCRIPTION
**Pull request by merbla**
Initial PR related to #593 for the following 

- Test AppVeyor Build
- Gather feed back from @serilog/core

Not ready to merge quite yet.

### CODE REVIEW

**Comment by merbla** [comment]
mmm is there a race condition in the tests?  @khellang how did you fix the last one?  just ran again?

[404](https://ci.appveyor.com/project/serilog/serilog/build/404) - :fire:
[403](https://ci.appveyor.com/project/serilog/serilog/build/403) - :ok_hand:
[402](https://ci.appveyor.com/project/serilog/serilog/build/402) - :fire:
[401](https://ci.appveyor.com/project/serilog/serilog/build/401) - :fire:

**Comment by khellang** [comment]
I think @nblumhardt got it passing by just running it again. It's flaky to say the least.

**Comment by nblumhardt** [comment]
I think we need another `ConfigureAwait(false)` in `PeriodicBatchingSink` somewhere; trying to find a chance to check out that theory.... :-)

**Comment by merbla** [comment]
ping @serilog/core.. review time pls

I think `src/Serilog.Abstractions/project.json` may get lighter once other cruft is moved in/out.

**Comment by nblumhardt** [comment]
Looks great @merbla - feels like the right direction :+1:

* `MessageTemplate.PostionalProperties` and `.NamedProperties` - exposing these turns the otherwise immutable `MessageTemplate` type into a subtly-mutable one; any alternatives here?
* `IMessageTemplateParser` - any way we can avoid exposing this/moving it to abstractions?

Cheers! :-)

**Comment by merbla** [comment]
Will have a look

File: src/Serilog.Abstractions/ILogger.cs [hunk 1]
```diff
@@ -48,7 +48,7 @@ public interface ILogger
         /// </summary>
         /// <returns>A logger that will enrich log events as specified.</returns>
         ILogger ForContext(string propertyName, object value, bool destructureObjects = false);
-
+
```
Comment (khellang) [inline src/Serilog.Abstractions/ILogger.cs]: :fire:

File: src/Serilog/project.json [hunk 2]
```diff
@@ -6,69 +6,72 @@
...
+        "System.Diagnostics.TraceSource": "4.0.0-beta-23516",
+        "System.IO": "4.0.11-beta-23516",
+        "System.IO.FileSystem": "4.0.1-beta-23516",
+        "System.Linq": "4.0.1-beta-23516",
+        "System.Text.RegularExpressions": "4.0.11-beta-23516",
+        "System.Threading.Thread": "4.0.0-beta-23516"
+      }
     }
+  },
+  "dependencies": {
```
Comment (khellang) [inline src/Serilog/project.json]: Can we move this above the `frameworks`? It's a bit hidden down here.

**Comment by khellang** [comment]
LGTM, just a couple of nits :shipit:

**Comment by merbla** [comment]
@nblumhardt @khellang thoughts on this approach?  

- Aiming to keep implementation internal etc.
- Is `IMessageTemplate` the way to go?

**Comment by nblumhardt** [comment]
Hey @merbla I'm not sure I'm on the same page, it's a bit tricky to piece together from the diff, sorry - guessing this is obvious but:

 * why do we need `IMessageTemplate`?
 * why does the message template parsing infrastructure need to go into Abstractions?

Sorry about the possibly clueless questions, thanks for pushing forward on this!

**Comment by merbla** [comment]
No worries @nblumhardt, really just feeling this one out.  I am sure there is more to come.

I have not gone for wholesale changes however we may need a few in that we have some interfaces that have dependencies on objects that rely on some internals of core

e.g. Currently the following is the landscape.

Ignoring `IMessageTemplate`, `LogEvent` has dependencies on interfaces and implementation that really should be in **Core**.  Similarly there are some interfaces that depend on `LogEvent` making this split troublesome.

I will have another think and ping you back when I have more.

**Comment by merbla** [comment]
@nblumhardt thoughts on an `ILogEvent`? Seems a little pointless as I can't see the need for other implementations.

**Comment by merbla** [comment]
Getting closer...

**Comment by nblumhardt** [comment]
Aha! I'm on the same page now @merbla - thanks for persevering.

I like where things have landed, but I am wondering if:

a) `SelfLog` belongs in Abstractions anyway, since most sinks use it, and
b) Instead of an interface, `IMessageTemplate` could be a class, called `MessageTemplate` and the optimised/caching version in core could be a subclass called `CachedMessageTemplate` and this would only be used by the few parts of the code requiring the optimised properties. E.g.:

```csharp
public class MessageTemplate
{
    readonly string _text;
    readonly MessageTemplateToken[] _tokens;

    public MessageTemplate(IEnumerable<MessageTemplateToken> tokens)
        : this(string.Join("", tokens), tokens)
    {
    }

    public MessageTemplate(string text, IEnumerable<MessageTemplateToken> tokens)
    {
        if (text == null) throw new ArgumentNullException(nameof(text));
        if (tokens == null) throw new ArgumentNullException(nameof(tokens));

        _text = text;
        _tokens = tokens.ToArray();
    }

    public string Text
    {
        get { return _text; }
    }

    public override string ToString()
    {
        return Text;
    }

    public IEnumerable<MessageTemplateToken> Tokens
    {
        get { return _tokens; }
    }

    public string Render(IPropertyDictionary properties, IFormatProvider formatProvider = null)
    {
        var writer = new StringWriter(formatProvider);
        Render(properties, writer, formatProvider);
        return writer.ToString();
    }

    public void Render(IPropertyDictionary properties, TextWriter output, IFormatProvider formatProvider = null)
    {
        foreach (var token in _tokens)
        {
            token.Render(properties, output, formatProvider);
        }
    }
}

class CachedMessageTemplate : MessageTemplate
{
    readonly PropertyToken[] _positionalProperties;
    readonly PropertyToken[] _namedProperties;

    public CachedMessageTemplate(string text, IEnumerable<MessageTemplateToken> tokens)
        : base(text, tokens)
    {
        var propertyTokens = tokens.OfType<PropertyToken>().ToArray();
        if (propertyTokens.Length != 0)
        {
            var allPositional = true;
            var anyPositional = false;
            foreach (var propertyToken in propertyTokens)
            {
                if (propertyToken.IsPositional)
                    anyPositional = true;
                else
                    allPositional = false;
            }

            if (allPositional)
            {
                _positionalProperties = propertyTokens;
            }
            else
            {
                if (anyPositional)
                    SelfLog.WriteLine("Message template is malformed: {0}", text);

                _namedProperties = propertyTokens;
            }
        }
    }

    public PropertyToken[] NamedProperties
    {
        get { return _namedProperties; }
    }

    public PropertyToken[] PositionalProperties
    {
        get { return _positionalProperties; }
    }
}
```

This would totally avoid the expensive interface/virtual dispatch, and potentially allow `CachedMessageTemplate` to remain internal (not sure but it's a possibility).

What do you think?

**Comment by nblumhardt** [comment]
@merbla nice diagrams BTW - what creates those? :-)

**Comment by merbla** [comment]
@nblumhardt will bang it out and let you know.

**Comment by nblumhardt** [comment]
@merbla I have a bit of time today, wondering if I might take on merging this and try out the `MessageTemplate`/`CachedMessageTemplate` transform in the process; what do you think? :-) Cheers!

**Comment by merbla** [comment]
Go for it!

**Comment by nblumhardt** [comment]
Have some time to spare, will pick this up and see if I can fix the build issue "on the way in" :-)

**Comment by nblumhardt** [comment]
This could be a really sucky time to come out with a comment like this ;-) ... but now after having dug into the changeset I'm wondering how much we're actually going to win when the dust settles.

To support the full scope of what client code and sinks will want to use, we'll at least need to bring across:

 * `Log` and `LogContext` and their dependencies (used by client code)
 * `TextToken` and `PropertyToken` (used by the console sinks and others)
 * `JsonFormatter` (used by many sinks)
 * `OutputProperties` and `MessageTemplateTextFormatter` (hazy recollection of sinks needing this)
 * `SelfLog` (used by many sinks)

At this point _Abstractions_ may not be the best name - there's a lot of implementation code mixed in there as well.

After doing that, the only things left in the _Serilog_ package are `LoggerConfiguration`, the core pipeline (changes rarely), enrichers, destructuring code, app settings support and the sinks.

Once we move the sinks, app settings support, and potentially the enrichers to their own packages, there won't be much to justify shipping _Serilog_ and _Serilog.Abstractions_ separately.

We're also left with some awkward structuring, like the complete infrastructure behind message templates being in _Abstractions_, but the parser itself sitting lonely in core.

Would I be crazy to suggest we put this PR on hold, and look at pulling out the sinks etc. as a first step, after all?

**Comment by merbla** [comment]
:hear_no_evil:  I think you are right.

Put this one on ice and start on the sinks and enrichers.

**Comment by nblumhardt** [comment]
@merbla I'll close this one for now since a re-take is probably going to be "from scratch" given all the work you've done on factoring since this PR was opened.

**Pull request closed without merging**